app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/triage/CommonTriageQuestionsController.scala

```diff
@@ -419,15 +420,86 @@ class CommonTriageQuestionsController @Inject() (
+                )
+              )
+            case _ if state.fold(_ => false, _.isAmendReturn)                                 =>
+              Ok(
+                previousReturnExistsWithSameCompletionDateAmendPage(
+                  value,
+                  state.fold(_.subscribedDetails, _.subscribedDetails).isATrust,
+                  backLink
+                )
+              )
+            case _                                                                            => Redirect(uk.gov.hmrc.cgtpropertydisposalsfrontend.controllers.routes.StartController.start())
+          }
+        }
       }
     }
 
+  private def withCompletionDate(state: Either[StartingNewDraftReturn, FillingOutReturn])(
+    f: CompletionDate => Future[Result]
+  ): Future[Result] =
+    state match {
+      case Left(value)  =>
+        value.newReturnTriageAnswers match {
+          case Left(value)  =>
+            value match {
+              case s: IncompleteMultipleDisposalsTriageAnswers                              =>
+                s.completionDate match {
+                  case Some(value) => f(value)
+                  case None        => Redirect(routes.MultipleDisposalsTriageController.completionDate())
+                }
+              case s: MultipleDisposalsTriageAnswers.CompleteMultipleDisposalsTriageAnswers => f(s.completionDate)
+            }
+          case Right(value) =>
+            value match {
+              case s: IncompleteSingleDisposalTriageAnswers                           =>
+                s.completionDate match {
+                  case Some(value) => f(value)
+                  case None        => Redirect(routes.SingleDisposalsTriageController.whenWasCompletionDate())
+                }
+              case s: SingleDisposalTriageAnswers.CompleteSingleDisposalTriageAnswers => f(s.completionDate)
+            }
+        }
+      case Right(value) =>
+        value.draftReturn.triageAnswers() match {
+          case Left(value)  =>
+            value match {
+              case m: IncompleteMultipleDisposalsTriageAnswers                              =>
+                m.completionDate match {
+                  case Some(value) => f(value)
+                  case None        => Redirect(routes.MultipleDisposalsTriageController.completionDate())
+                }
+              case m: MultipleDisposalsTriageAnswers.CompleteMultipleDisposalsTriageAnswers => f(m.completionDate)
+            }
+          case Right(value) =>
+            value match {
+              case s: IncompleteSingleDisposalTriageAnswers                           =>
+                s.completionDate match {
+                  case Some(value) => f(value)
+                  case None        => Redirect(routes.SingleDisposalsTriageController.whenWasCompletionDate())
+                }
+              case s: SingleDisposalTriageAnswers.CompleteSingleDisposalTriageAnswers => f(s.completionDate)
+            }
+        }
+    }
```

Comment:
there should be some way to factor out the common code in this function by perhaps having  local `def` inside it